### DISCUSSION TITLE
How to use amend locally

### DISCUSSION THREAD

**Discussion by lonix1** [jesseduffield/lazygit#4025] [Q&A] [answered]
New user here. Hope someone can explain what I'm doing incorrectly.

I want to do an amend, so:
- I stage some changes
- press escape to return to the Files panel
- press shift-A, it asks "Are you sure you want to amend last commit? Afterwards, you can..."
- press Enter

It shows an error:
> Git command failed. Check command log for details (open with @)

I press @ then "Focus command log" and it shows
> Git output:
> ssh: Could not resolve hostname example.com: Name or service not known
> fatal: Could not read from remote repository.

I disconnected the network to avoid making mistakes while learning to use lazygit. Glad I did!

I just want to amend the last commit, not push. Why is it trying to reach the remote, and how do I prevent that?

lazygit 0.44.1, linux

**Comment by mark2185** [comment]
It _should_ work offline, the only thing I could think of that could fail offline is the auto fetch (which is enabled by default), but that wouldn't happen at the exact moment when you're amending.

A few things to help narrow this down:
- does `git commit --amend --no-edit` in the terminal work okay?
- do you have any `pre-commit` hooks in the repo?
- could you please post your `config.yml`?
- does anything related to `git amend` appear in the log window?

Also you can try it in a new repo that doesn't have an upstream, that way you don't have to worry about it being pushed anywhere:
```
$> git init test-repo
$> cd test-repo
$> # do whatever you want
```

**Reply by lonix1** [reply]
> could you please post your `config.yml`?

I found one (default I guess) in `~/.config/lazygit/config.yml`, which is empty.

> do you have any `pre-commit` hooks in the repo?

I did but it had nothing to do with fetching (I think...); I removed it anyway and it worked!

The hook was copied from [here](https://codeinthehole.com/tips/tips-for-using-a-git-pre-commit-hook/#search-yo-code-for-debugging-code), to look for "TODO" comments in code (and prevent such commits):
```sh
#!/bin/sh
PATTERN_FILES='\.(cs|cshtml|csproj|css|js|scss|sh|ts|xml|yml)?$'
FORBIDDEN_TEXT='TODO'
git diff --cached --name-only | \
  grep -E "$PATTERN_FILES" | \
  GREP_COLOR='4;5;37;41' xargs grep --color --with-filename -n "$FORBIDDEN_TEXT" \
  && echo 'ERROR: Found forbidden texts in staged files; remove them then try again' \
  && exit 1
```

I don't think that contacts the remote. However, when running that script directly, I get an error (must be some compat problem after upgrading to a newer bash):
```sh
grep: warning: GREP_COLOR='4;5;37;41' is deprecated; use GREP_COLORS='mt=4;5;37;41'
```

My suspicion is due to that error, lazygit fails (probably a good idea), but its error message is misleading.

I fixed the script, tried again, and received the same error though. Very weird! I wonder why lazygit chokes on that hook.

Does lazygit work with pre-commit hooks?

**Reply by mark2185** [reply, marked as answer]
Does the script reach the `exit 1` part? If it does, then that's probably why `lazygit` considers it an error, because the exit code is non-zero.

**Reply by lonix1** [reply]
The script was not returning a proper exit code for non-error, so I fixed it. Just for reference:
```sh
#!/bin/sh
PATTERN_FILES='\.(cs|cshtml|csproj|css|js|scss|sh|ts|xml|yml)?$'
FORBIDDEN_TEXT='TODO'
git diff --cached --name-only | \
  sed -n -E "/$PATTERN_FILES/p" | \
  GREP_COLORS='4;5;37;41' xargs grep --color --with-filename -n "$FORBIDDEN_TEXT" \
  && echo 'ERROR: Found forbidden texts in staged files; remove them then try again' && exit 1 \
  || exit 0
```
Now it works with lazygit.

The original confusion persists though... if the pre-commit hook fails (exit code != 0) then it complains about the remote. It should probably complain about the hook.

Thanks for your help, enjoying lazygit so far!

**Reply by mark2185** [reply]
> The original confusion persists though... if the pre-commit hook fails (exit code != 0) then it complains about the remote.

Yeah, that _is_ odd. Does that also happen while you're _not_ offline? I'm wondering if maybe the failing hook just flushes the buffer that had the error message from the background fetch failing.

**Reply by lonix1** [reply]
Yes even when online, that was the error message.
